### PR TITLE
Fix DNS resolution in ephemeral guests

### PR DESCRIPTION
**Pull request by gursewak1997**
Configure QEMU user-mode networking to use host DNS servers from /etc/resolv.conf instead of the default 10.0.2.3, which doesn't work when QEMU runs inside containers.

### CODE REVIEW

File: crates/kit/src/run_ephemeral.rs [hunk 1]
```diff
@@ -1229,13 +1246,44 @@ Options=
...
+
     if opts.common.ssh_keygen {
         qemu_config.enable_ssh_access(None); // Use default port 2222
         debug!("Enabled SSH port forwarding: host port 2222 -> guest port 22");
 
         // We need to extract the public key from the SSH credential to inject it via SMBIOS
         // For now, the credential is already being passed via kernel cmdline
         // TODO: Add proper SMBIOS credential injection if needed
+    } else {
+        // Even without SSH, ensure DNS is configured if we have DNS servers
```
Comment (cgwalters) [inline crates/kit/src/run_ephemeral.rs]: I don't understand why we need to do this twice

File: crates/kit/src/qemu.rs [hunk 2]
```diff
@@ -22,6 +22,38 @@ use nix::sys::socket::{accept, bind, getsockname, socket, AddressFamily, SockFla
 use tracing::{debug, trace, warn};
 use vsock::VsockAddr;
 
+/// Read DNS servers from /etc/resolv.conf
```
Comment (cgwalters) [inline crates/kit/src/qemu.rs]: While this is a pretty simple file format, it's a bit unfortunate for us to carry another one. But I guess this is really tied up in the "is there a better (different) way for us to solve this" larger issue.

File: crates/kit/src/qemu.rs [hunk 3]
```diff
@@ -22,6 +22,38 @@ use nix::sys::socket::{accept, bind, getsockname, socket, AddressFamily, SockFla
...
+pub fn read_dns_servers() -> Option<Vec<String>> {
+    let resolv_conf = match std::fs::read_to_string("/etc/resolv.conf") {
+        Ok(content) => content,
+        Err(e) => {
+            debug!("Failed to read /etc/resolv.conf: {}", e);
+            return None;
+        }
+    };
+
+    let mut dns_servers = Vec::new();
```
Comment (cgwalters) [inline crates/kit/src/qemu.rs]: General best practice is to split parsers from data reading - this is obviously a trivial parser, but it'd be easy to unit test the parser if it accepted a `&str` or so, then `read_dns_servers` just reads `/etc/resolv.conf` and passes the string to it

File: crates/integration-tests/src/tests/run_ephemeral_ssh.rs [hunk 4]
```diff
@@ -358,3 +358,67 @@ fn test_run_ephemeral_ssh_broken_image_cleanup() -> Result<()> {
...
+    let dns_output = run_bcvk(&[
+        "ephemeral",
+        "run-ssh",
+        "--label",
+        INTEGRATION_TEST_LABEL,
+        &get_test_image(),
+        "--",
+        "/bin/sh",
+        "-c",
+        r#"getent hosts google.com || ping -c1 -W2 google.com | head -1"#,
```
Comment (cgwalters) [inline crates/integration-tests/src/tests/run_ephemeral_ssh.rs]: I'd say use quay.io since we require that service to exist for other things

And it seems cleaner to just require `getent hosts` to work, why the fallback to ping?

File: crates/integration-tests/src/tests/run_ephemeral_ssh.rs [hunk 5]
```diff
@@ -358,3 +358,67 @@ fn test_run_ephemeral_ssh_broken_image_cleanup() -> Result<()> {
...
+
+    let out = dns_output.stdout;
+    let resolved = out.contains("google.com")
+        || out.contains("PING")
+        || out.contains("2001:")
+        || out.contains("142.")
+        || out.contains("104.")
+        || out.contains("108.")
+        || out.contains("172.217.")
+        || out.contains("216.58.");
```
Comment (cgwalters) [inline crates/integration-tests/src/tests/run_ephemeral_ssh.rs]: This list of hardcoded IP addresses seems a bit weird...why not just require `getent hosts` to work?

Or actually building on the above, how about just `podman pull quay.io/fedora/fedora` or so for a more realistic networking test?

File: crates/integration-tests/src/tests/run_ephemeral_ssh.rs [hunk 6]
```diff
@@ -358,3 +358,68 @@ fn test_run_ephemeral_ssh_broken_image_cleanup() -> Result<()> {
...
+        "run-ssh",
+        "--label",
+        INTEGRATION_TEST_LABEL,
+        &get_test_image(),
+        "--",
+        "/bin/sh",
+        "-c",
+        r#"
+        for i in $(seq 1 30); do
+            ip -4 addr show | grep -q "inet " && break
```
Comment (cgwalters) [inline crates/integration-tests/src/tests/run_ephemeral_ssh.rs]: Because ssh requires IP networking, I can't think of a scenario in which ssh would work but this could somehow fail. This is likely just dead code.

Comment (gursewak1997) [reply crates/integration-tests/src/tests/run_ephemeral_ssh.rs]: I see. Updated and removed the dead code there.

**Comment by gursewak1997** [comment]
Reworking this PR as I think it will require a few more changes.

**Comment by gursewak1997** [comment]
This one's just timing out on test_to_disk_for_image_quay_io_centos_bootc_centos_bootc_stream9 test but should be good to go though

File: crates/kit/src/run_ephemeral.rs [hunk 7]
```diff
@@ -1229,6 +1321,31 @@ Options=
...
+        for server in &dns_servers {
+            resolv_content.push_str(&format!("nameserver {}\n", server));
+        }
+
+        std::fs::write("/etc/resolv.conf", &resolv_content)
+            .context("Failed to write /etc/resolv.conf for QEMU slirp")?;
+
+        debug!("Configured DNS servers for QEMU slirp: {:?}", dns_servers);
+
+        // Verify the file was written correctly
```
Comment (cgwalters) [inline crates/kit/src/run_ephemeral.rs]: Seems strange to double check the result of a core library function `std::fs::write`

File: crates/kit/src/run_ephemeral.rs [hunk 8]
```diff
@@ -1229,6 +1321,31 @@ Options=
...
+    // which contains unreachable bridge DNS servers (e.g., 169.254.1.1, 10.x.y.z).
+    // Solution: Write public DNS servers to /etc/resolv.conf in the bwrap namespace.
+    // Safe because we're in an ephemeral container that will be destroyed.
+    if let Some(dns_servers) = opts.host_dns_servers.clone() {
+        let mut resolv_content = String::new();
+        for server in &dns_servers {
+            resolv_content.push_str(&format!("nameserver {}\n", server));
+        }
+
+        std::fs::write("/etc/resolv.conf", &resolv_content)
```
Comment (cgwalters) [inline crates/kit/src/run_ephemeral.rs]: There's some subtleties about writing `/etc/resolv.conf`. For example, it might be a symlink - and by default with podman it's actually a bind mount to a tmpfs:

```
$ podman run --rm -ti quay.io/centos-bootc/centos-bootc:stream10 findmnt /etc/resolv.conf
TARGET           SOURCE                                                                                                                      FSTYPE OPTIONS
/etc/resolv.conf tmpfs[/containers/overlay-containers/9e6bf779db5f95efb19cd1d4f461cbdfba1ff0a276f7f13a45457f222e9d7c06/userdata/resolv.conf] tmpfs  rw,nosuid,nodev,relatime,seclabel,size=6560416k,nr_inodes=1640104,mode=700,uid=1002,gid=1002,inode64
$
```

Overwriting it this way will work, and it appears that podman at least doesn't try to re-synchronize it as the host content changes.

Hum...on this topic, it looks like we could actually override using `podman run --dns`?

File: crates/kit/src/run_ephemeral.rs [hunk 9]
```diff
@@ -283,6 +283,82 @@ pub struct RunEphemeralOpts {
...
+                // We'll fall back to public DNS (8.8.8.8, 1.1.1.1) which is more reliable.
+                let filtered_servers: Vec<String> = dns_servers
+                    .into_iter()
+                    .filter(|s| {
+                        // Parse as IPv4 address
+                        if let Ok(ip) = s.parse::<std::net::Ipv4Addr>() {
+                            // Reject loopback, link-local, and private addresses
+                            !ip.is_loopback() && !ip.is_link_local() && !ip.is_private()
+                        } else {
+                            false // Reject IPv6 and invalid addresses
```
Comment (cgwalters) [inline crates/kit/src/run_ephemeral.rs]: Why filter out ipv6?

File: crates/integration-tests/src/tests/run_ephemeral_ssh.rs [hunk 10]
```diff
@@ -358,3 +358,67 @@ fn test_run_ephemeral_ssh_broken_image_cleanup() -> Result<()> {
     Ok(())
 }
 integration_test!(test_run_ephemeral_ssh_broken_image_cleanup);
+
+/// Test ephemeral VM network and DNS
+///
+/// Verifies that ephemeral bootc VMs can access the network and resolve DNS correctly.
+/// Uses HTTP request to quay.io to test both DNS resolution and network connectivity.
+fn test_run_ephemeral_dns_resolution() -> Result<()> {
+    // Wait for network interface to be ready
```
Comment (cgwalters) [inline crates/integration-tests/src/tests/run_ephemeral_ssh.rs]: Looks like this came back in, I still think it's unnecessary

Comment (gursewak1997) [reply crates/integration-tests/src/tests/run_ephemeral_ssh.rs]: ah yes, the test did start failing and I re-added everything. Since it's changed fully, I can def remove this now

**Review by cgwalters — APPROVED** [review]

**Comment by gursewak1997** [comment]
> So I get this from every run now:
> 
> > WARN No usable DNS servers found in system configuration, falling back to public DNS (8.8.8.8, 1.1.1.1). This may not work in air-gapped environments.
> 
> Looks like what's happening is when I'm on my work VPN the DNS is switched over to 10.x which matches `is_private`. It's not totally clear to me why we'd need to filter out private IP addresses.
> 

A fix for now: https://github.com/bootc-dev/bcvk/pull/182

File: crates/kit/src/run_ephemeral.rs [hunk 11]
```diff
@@ -283,6 +283,87 @@ pub struct RunEphemeralOpts {
...
+
+    /// Host DNS servers (read on host, configured via podman --dns flags)
+    /// Not a CLI option - populated automatically from host's /etc/resolv.conf
+    #[clap(skip)]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub host_dns_servers: Option<Vec<String>>,
+}
+
+/// Parse DNS servers from resolv.conf format content
+fn parse_resolv_conf(content: &str) -> Vec<String> {
```
Comment (cgwalters) [inline crates/kit/src/run_ephemeral.rs]: AI: **Suggestion**: The `parse_resolv_conf` function is a good candidate for unit tests. It parses a structured format and has edge cases (empty lines, whitespace, missing servers) that would benefit from table-driven tests verifying correct parsing of typical and edge-case resolv.conf contents.

File: crates/kit/src/run_ephemeral.rs [hunk 12]
```diff
@@ -283,6 +283,87 @@ pub struct RunEphemeralOpts {
...
+/// Read DNS servers from host's resolv.conf
+/// Returns a vector of DNS server IP addresses, or None if unable to read/parse
+///
+/// For systemd-resolved systems, reads from /run/systemd/resolve/resolv.conf
+/// which contains actual upstream DNS servers, not the stub resolver (127.0.0.53).
+/// Falls back to /etc/resolv.conf for non-systemd-resolved systems.
+fn read_host_dns_servers() -> Option<Vec<String>> {
+    // Try systemd-resolved's upstream DNS file first
+    // This avoids reading 127.0.0.53 (stub resolver) from /etc/resolv.conf
+    let paths = [
```
Comment (cgwalters) [inline crates/kit/src/run_ephemeral.rs]: AI: **Suggestion**: Consider extracting the systemd-resolved and fallback paths into constants, e.g. `const SYSTEMD_RESOLVED_UPSTREAM: &str = "/run/systemd/resolve/resolv.conf";` This makes them easier to find and document.

File: crates/kit/src/run_ephemeral.rs [hunk 13]
```diff
@@ -283,6 +283,87 @@ pub struct RunEphemeralOpts {
...
+    ];
+
+    for path in &paths {
+        match std::fs::read_to_string(path) {
+            Ok(content) => {
+                let dns_servers = parse_resolv_conf(&content);
+
+                // Filter out localhost, link-local, and private network addresses
+                // QEMU runs in user networking mode (slirp) inside a container, which cannot
+                // reach private network addresses (10.x.x.x, 172.16-31.x.x, 192.168.x.x for IPv4,
```
Comment (cgwalters) [inline crates/kit/src/run_ephemeral.rs:331]: AI: *Important*: The filtering of private network addresses (`!ip.is_private()` for IPv4, ULA check for IPv6) will break DNS for many enterprise/VPN users whose DNS servers are on private networks. Consider only filtering loopback and link-local. (Note: This was fixed in follow-up commit 8f8d301)

File: crates/integration-tests/src/tests/run_ephemeral_ssh.rs [hunk 14]
```diff
@@ -358,3 +358,43 @@ fn test_run_ephemeral_ssh_broken_image_cleanup() -> Result<()> {
...
+        "run-ssh",
+        "--label",
+        INTEGRATION_TEST_LABEL,
+        &get_test_image(),
+        "--",
+        "/bin/sh",
+        "-c",
+        r#"
+        if command -v curl >/dev/null 2>&1; then
+            curl -sS --max-time 10 https://quay.io/v2/ >/dev/null
```
Comment (cgwalters) [inline crates/integration-tests/src/tests/run_ephemeral_ssh.rs:381]: AI: (low) This test requires external network access to quay.io, which may fail in air-gapped CI environments or during quay.io outages. Consider documenting this as an integration test requirement, or adding a skip mechanism for offline testing.

File: crates/kit/src/run_ephemeral.rs [hunk 15]
```diff
@@ -283,6 +283,87 @@ pub struct RunEphemeralOpts {
...
     pub kernel_args: Vec<String>,
+
+    /// Host DNS servers (read on host, configured via podman --dns flags)
+    /// Not a CLI option - populated automatically from host's /etc/resolv.conf
+    #[clap(skip)]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub host_dns_servers: Option<Vec<String>>,
+}
+
+/// Parse DNS servers from resolv.conf format content
```
Comment (cgwalters) [inline crates/kit/src/run_ephemeral.rs:294]: AI: Good structure - having `parse_resolv_conf` take `&str` makes it unit-testable without filesystem dependencies. Consider adding unit tests for edge cases (empty file, malformed lines, mixed IPv4/IPv6).
